kaitai/stream_test.go

```diff
@@ -74,6 +75,99 @@ func TestStream_Size(t *testing.T) {
 	}
 }
 
+type artificialError struct{}
+
+func (e artificialError) Error() string {
+	return "artificial error when seeking with io.SeekCurrent after seeking to end"
+}
+
+type failingReader struct {
+	pos      int64
+	mustFail func(fr failingReader, offset int64, whence int) bool
+}
+
+func (fr *failingReader) Read(p []byte) (n int, err error) { return 0, nil }
+func (fr *failingReader) Seek(offset int64, whence int) (int64, error) {
+	if fr.mustFail(*fr, offset, whence) {
+		return 0, artificialError{}
+	}
+
+	switch {
+	case whence == io.SeekCurrent:
+		return fr.pos, nil
+	case whence == io.SeekStart:
+		fr.pos = offset
+	default: // whence == io.SeekEnd
+		fr.pos = -1
+	}
+
+	return fr.pos, nil
+}
+
+// No regression test for issue #26
+func TestErrorHandlingInStream_Size(t *testing.T) {
+	tests := map[string]struct {
+		initialPos       int64
+		failingCondition func(fr failingReader, offset int64, whence int) bool
+		errorCheck       func(err error) bool
+		wantFinalPos     int64
+	}{
+		"fails to get initial position": {
+			initialPos: 5,
+			failingCondition: func(fr failingReader, offset int64, whence int) bool {
+				return whence == io.SeekCurrent && offset == 0
+			},
+			errorCheck: func(err error) bool {
+				_, ok := err.(artificialError)
+				return ok
+			},
```

Comment:
I don't quite get why this check function is declared for each case individually. All cases use `failingReader`:

https://github.com/kaitai-io/kaitai_struct_go_runtime/blob/3e00f498f7d4de439099098e7d62cc84b541d005/kaitai/stream_test.go#L151-L153

which in turn **always** returns `artificialError` if `failingCondition` (also known as `mustFail`) is true:

https://github.com/kaitai-io/kaitai_struct_go_runtime/blob/3e00f498f7d4de439099098e7d62cc84b541d005/kaitai/stream_test.go#L91-L93

So I'd rather replace [`stream_test.go:159-161`](https://github.com/kaitai-io/kaitai_struct_go_runtime/blob/3e00f498f7d4de439099098e7d62cc84b541d005/kaitai/stream_test.go#L159-L161) with the previous version:

https://github.com/kaitai-io/kaitai_struct_go_runtime/blob/b841c18ae8609296514002ea13ff475e81eb9769/kaitai/stream_test.go#L114-L117

Reply:
Not all cases return `artificialError`: when the deferred seek fails it returns an `artificialError` **wrapped** into a `fmt.wrapError`.

A more explicit definition could be:
```go
func TestErrorHandlingInStream_Size(t *testing.T) {
	isArtificialError := func(err error) bool {
		_, ok := err.(artificialError)
		return ok
	}

	isWrappedArtificialError := func(err error) bool {
		_, ok := errors.Unwrap(err).(artificialError)
		return ok
	}

	tests := map[string]struct {
		initialPos       int64
		failingCondition func(fr failingReader, offset int64, whence int) bool
		errorCheck       func(err error) bool
		wantFinalPos     int64
	}{
		"fails to get initial position": {
			initialPos: 5,
			failingCondition: func(fr failingReader, offset int64, whence int) bool {
				return whence == io.SeekCurrent && offset == 0
			},
			errorCheck:   isArtificialError,
			wantFinalPos: 5,
		},
		"seek to the end fails": {
			initialPos: 5,
			failingCondition: func(fr failingReader, offset int64, whence int) bool {
				return whence == io.SeekEnd
			},
			errorCheck:   isArtificialError,
			wantFinalPos: 5,
		},
		"deferred seek to the initial pos fails": {
			initialPos: 5,
			failingCondition: func(fr failingReader, offset int64, whence int) bool {
				return whence == io.SeekStart && fr.pos == -1
			},
			errorCheck:   isWrappedArtificialError,
			wantFinalPos: -1,
		},
	}
```
Anyway, you can go ahead, pick the version you like the most and modify the PR consequently.

Reply:
> Not all cases return `artificialError`: when the deferred seek fails it returns an `artificialError` **wrapped** into a `fmt.wrapError`.

Oh, thanks, I didn't notice that. I didn't see this little exclamation mark in `!ok` here:

https://github.com/kaitai-io/kaitai_struct_go_runtime/blob/fc91f260c0085f65c6e7f0e40bc4a9da87bf526f/kaitai/stream_test.go#L142-L145

But it turns out that the linter knows about the problem that wrapped errors cannot be just type asserted, and it suggested a solution:

```console
$ ./golangci-lint-1.35.2-windows-amd64/golangci-lint run kaitai --no-config -E errorlint
kaitai\stream_test.go:121:14: type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors (errorlint)
                                _, ok := err.(artificialError)
                                         ^
```

So I believe that this silver bullet [`errors.As`](https://golang.org/pkg/errors/#As) should work for everything:

```diff
+	var wantErr artificialError
-	_, isArtificialErr := err.(artificialError)
+	_, isArtificialErr := errors.As(err, &wantErr)
 	if !isArtificialErr {
-		t.Fatalf("Expected error of type %T, got one of type %T", artificialError{}, err)
+		t.Fatalf("Expected error of type %T, got one of type %T", wantErr, err)
 	}
```

See https://play.golang.org/p/1FN6BVmqRWV for a demo.